src/ast/passes/semantic_analyser.cpp

```diff
@@ -514,7 +519,51 @@ void SemanticAnalyser::visit(Call &call)
       auto &expr = (*call.vargs)[i];
       func_arg_idx_ = i;
 
-      expr->accept(*this);
+      // Handling offsetof() 2nd argument
+      // Resolve the second offsetof() parameter as early as possible,
+      // otherwise, we need to do too much adaptation work in the semantic
+      // analyzer stage. For example:
+      //
+      //  struct foo {
+      //    int offsetof;   > offsetof(struct foo, offsetof)
+      //    int kstack;     > offsetof(struct foo, kstack)
+      //    int ctx;        > offsetof(struct foo, ctx)
+      //    int comm;       > offsetof(struct foo, comm)
+      //  }
+      if (func_ == "offsetof" && func_arg_idx_ == 1)
+      {
+        auto fc = dynamic_cast<FieldAccess *>(call.vargs->at(1));
+        auto binop = dynamic_cast<Binop *>(call.vargs->at(1));
+
+        // TODO: Only support single-member form for now.
+        // a.b in offsetof(struct Foo, a.b) is illegal
+        if (fc != nullptr)
+        {
+          LOG(ERROR, call.loc, err_)
+              << call.func << "() unsupport subfield format.";
+        }
+        // a+b in offsetof(struct Foo, a+b) is illegal
+        else if (binop != nullptr)
+        {
+          LOG(ERROR, call.loc, err_)
+              << call.func << "() unsupport expression format.";
+        }
+        else
+        {
+          auto &field_arg = *call.vargs->at(1);
+          String &field_str = static_cast<String &>(field_arg);
```

Comment:
Hmm, not sure this works. The second argument doesn't _have_ to be a string -- this might just segfault if it's something else. In fact I think it cannot be a string -- a string would be `"field"` (note the quotes). This might only be working b/c `String` has a similar layout to `Identifier` which is what you actually want here. And I think this needs to be a `dynamic_cast` to `Identifier` similar to the checks above

Reply:
Fix it.